bzzeth/bzzeth.go

```diff
@@ -326,6 +372,256 @@ func newChunk(data []byte) chunk.Chunk {
+			break DELIVERY
+		}
+		select {
+		case trigger <- batches: // signal that we are ready for another batch
+		case <-ctx.Done():
+			break DELIVERY
+		}
+	}
+	p.logger.Debug("bzzeth.handleGetBlockHeaders", "id", msg.Rid, "total", total, "delivered", deliveredCnt, "err", err)
+	if err == nil && deliveredCnt < total { // if there was no send error and we deliver less than requested
+		err := p.Send(ctx, &BlockHeaders{Rid: uint32(msg.Rid)}) // it is prudent to send an empty BlockHeaders message
+		if err != nil {
+			p.logger.Error("could not send empty BlockHeader")
+		}
+	}
+	p.logger.Debug("bzzeth.handleGetBlockHeaders: sent all headers", "id", msg.Rid)
+}
+
+var batchWait = 100 * time.Millisecond // time to wait for collecting headers in a batch
+var minBatchSize = 1                   // minimum headers in a batch
+
+// readToBatches reads items from an input channel into a buffer and
+// sends non-empty buffers on a channel read from the out
+func readToBatches(in chan []byte, out chan chan [][]byte) {
+	var buffer [][]byte
+	var trigger chan chan [][]byte
+BATCH:
+	for {
+		select {
+		case batches := <-trigger: // new batch channel available
+			if batches == nil { // terminate if batches channel is closed, no more batches accepted
+				return
+			}
+			batches <- buffer // otherwise write buffer into batch channel
+			if in == nil {    // terminate if in channel is already closed, sent last batch
+				return
+			}
+			buffer = nil  // otherwise start new buffer
+			trigger = nil // block this case: disallow new batches until enough in buffer
+
+		case item, more := <-in: // reading input
+
+			if !more {
+				in = nil       // block this case: disallow read from closed channel
+				continue BATCH // wait till last batch can send
+			}
+			// otherwise collect item in buffer
+			buffer = append(buffer, item)
+		default:
+			if len(buffer) >= minBatchSize { // if buffer is not empty
+				trigger = out  // allow sending batch
+				continue BATCH // wait till next batch can send
+			}
+			time.Sleep(batchWait) // otherwise wait and continue
+		}
+	}
+}
+
+// getBlockHeaderBzz retrieves a block header by its hash from swarm
+func (b *BzzEth) getBlockHeaderBzz(ctx context.Context, hash []byte) ([]byte, error) {
+	req := &storage.Request{
+		Addr:   hash,
+		Origin: b.netStore.LocalID,
+	}
+	chnk, err := b.netStore.Get(ctx, chunk.ModeGetRequest, req)
+	if err != nil {
+		return nil, err
+	}
+	return chnk.Data(), nil
+}
+
+// requestAll requests each hash and channel
+func (b *BzzEth) requestAll(ctx context.Context, deliveries chan []byte, hashes [][]byte, rcvdPeer *Peer) {
+	ctx, cancel := context.WithTimeout(ctx, timeouts.FetcherGlobalTimeout)
+	defer cancel()
+
+	// missingHeaders collects hashes of headers not found within swarm
+	// ie., the hashes to request from the eth full nodes
+	missingHeaders := make(chan []byte)
+	defer close(missingHeaders)
+	var wg sync.WaitGroup
+
+BZZ:
+	for _, h := range hashes {
+		hdr := make([]byte, len(h))
```

Comment:
Why is a copy needed? Could the has address be just passed to the goroutine bellow, or we intend to mutate it?

Reply:
I now pass the header in to the goroutine.. otherwise the same buffer gets access in all the go routines..